### DISCUSSION TITLE
why not just use OnValidate and use ??=

### DISCUSSION THREAD

**Discussion by IAFahim** [apkd/Medicine#7] [General]
Example like

```csharp
using UnityEngine;

namespace Data.Classes.Group
{
    public class RigidbodyColliderGroup : MonoBehaviour
    {
        public Rigidbody rb;
        public Collider col;

        protected virtual void OnValidate()
        {
            rb ??= GetComponent<Rigidbody>();
            col ??= GetComponent<Collider>();
        }
        
        public virtual void DisableComponents()
        {
            col.enabled = false;
            rb.isKinematic = true;
        }
        
        public virtual void EnableComponents()
        {
            col.enabled = true;
            rb.isKinematic = false;
        }
    }
}
```

**Comment by bulleador** [comment]
1. It wouldn't scale well if your component grows. You have to modify `EnableComponents`, `DisableComponents` and `OnValidate` after adding or removing a dependency.
2. Boilerplate. I think that the main purpose of this package is to reduce the amount of code you have to write and maintain.

**Reply by IAFahim** [reply]
Wow, you are fast.

Are you doing those in editor or run time that can be a question right. `OnValidate` is on unity editor and you get those reference via Serialization. And the game doesn't have to look for those components at run time of the game.

Example of which you show at on `Awake`. It's kinda bad way right!
I don't think I know unity more then you do as you wrote a whole package on it. I wanted to know is it any more performant doing it your way, other than having to write less code.

And I do find it sexy looking at your one liner instead of writing so much code. Asking if I have give away performance, that's all.

**Reply by IAFahim** [reply]
So if I use `OnValidate` it attachs those `GameObjects` for me to those referencees in editor before I hit play. So the game startup quite quickly, right?

**Reply by bulleador** [reply]
Oh, just to clarify things, @apkd is the creator of Medicine, not me :)

Yes, AFAIK the injecton happens on Awake but tbh I believe that the overhead introduced by GetComponent calls is negligible if you cache references and avoid GetComponent in Update.

To answer your question, I'm certain (although i've done absolutely no profiling to test this) that getting your dependencies in OnValidate is more efficient but the boilerplate code you have to maintain might not be worth the close-to-zero performance improvement.

Also, what if you want to add a component at runtime? OnValidate will not run outside the Editor.

**Reply by IAFahim** [reply]
Wait you are not the maintainer and still you are this quick. Wow. Quite impressive. I guess I really really really like the inspector in unity. I am like full anti-singleton and he is like full pro-singleton. He would really really like this library. I can't handle the fact that I can't see what stuff it's going to use in inspector. Thank you so much @bulleador
